backends/file/client.go

```diff
@@ -3,67 +3,129 @@ package file
 }
 
-func NewFileClient(filepath string) (*Client, error) {
-	return &Client{filepath}, nil
+type ResultError struct {
+	response uint64
+	err      error
 }
 
-func (c *Client) GetValues(keys []string) (map[string]string, error) {
-	yamlMap := make(map[interface{}]interface{})
-	vars := make(map[string]string)
+func NewFileClient(filepath []string, filter string) (*Client, error) {
+	return &Client{filepath: filepath, filter: filter}, nil
+}
 
-	data, err := ioutil.ReadFile(c.filepath)
+func readFile(path string, vars map[string]string) error {
+	yamlMap := make(map[interface{}]interface{})
+	data, err := ioutil.ReadFile(path)
 	if err != nil {
-		return vars, err
+		return err
 	}
+
 	err = yaml.Unmarshal(data, &yamlMap)
 	if err != nil {
-		return vars, err
+		return err
 	}
 
-	nodeWalk(yamlMap, "", vars)
-	log.Debug(fmt.Sprintf("Key Map: %#v", vars))
+	err = nodeWalk(yamlMap, "/", vars)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
+func (c *Client) GetValues(keys []string) (map[string]string, error) {
+	vars := make(map[string]string)
+	var filePaths []string
+	for _, path := range c.filepath {
+		p, _, err := util.Lookup(path, c.filter)
+		if err != nil {
+			return nil, err
+		}
+		filePaths = append(filePaths, p...)
+	}
+
+	for _, path := range filePaths {
+		readFile(path, vars)
+	}
+
+VarsLoop:
+	for k, _ := range vars {
+		for _, key := range keys {
+			if strings.HasPrefix(k, key) {
+				continue VarsLoop
+			}
+		}
+		delete(vars, k)
+	}
+	log.Debug(fmt.Sprintf("Key Map: %#v", vars))
 	return vars, nil
 }
 
 // nodeWalk recursively descends nodes, updating vars.
-func nodeWalk(node map[interface{}]interface{}, key string, vars map[string]string) error {
-	for k, v := range node {
-		key := key + "/" + k.(string)
-
-		switch v.(type) {
-		case map[interface{}]interface{}:
-			nodeWalk(v.(map[interface{}]interface{}), key, vars)
-		case []interface{}:
-			for _, j := range v.([]interface{}) {
-				switch j.(type) {
-				case map[interface{}]interface{}:
-					nodeWalk(j.(map[interface{}]interface{}), key, vars)
-				case string:
-					vars[key+"/"+j.(string)] = ""
-				}
-			}
-		case string:
-			vars[key] = v.(string)
+func nodeWalk(node interface{}, key string, vars map[string]string) error {
+	switch node.(type) {
+	case []interface{}:
+		for i, j := range node.([]interface{}) {
+			key := path.Join(key, strconv.Itoa(i))
+			nodeWalk(j, key, vars)
 		}
+	case map[interface{}]interface{}:
+		for k, v := range node.(map[interface{}]interface{}) {
+			key := path.Join(key, k.(string))
+			nodeWalk(v, key, vars)
+		}
+	case string:
+		vars[key] = node.(string)
+	case int:
+		vars[key] = strconv.Itoa(node.(int))
+	case bool:
+		vars[key] = strconv.FormatBool(node.(bool))
+	case float64:
+		vars[key] = strconv.FormatFloat(node.(float64), 'f', -1, 64)
 	}
 	return nil
 }
 
+func (c *Client) watchChanges(watcher *fsnotify.Watcher, stopChan chan bool) ResultError {
+	outputChannel := make(chan ResultError)
+	go func() error {
```

Comment:
This one will cause goroutine leak. `WatchPrefix` function is actually called in a loop in the main function. `WatchPrefix` should block until it receives an event and exit afterward.